### DISCUSSION TITLE
Issue in Retaining the Origninal Shape of Adjacency Matrix after Conversion from Dense to Sparse

### DISCUSSION THREAD

**Discussion by srikesh-07** [pyg-team/pytorch_geometric#9149] [Q&A]
**Hello Community,**

Thank you for intending to help me.

 I am working on a custom bipartite dataset that seems good until I meet this issue. The issue is while reconstructing the edge indices from dense matrix of  `Data` instances by the `DataLoader`

```
BipartiteGraph(x_constraints=[180, 1], edge_index=[2, 8130], edge_attr=[8130, 1], x_variables=[255, 9], num_nodes=435)
BipartiteGraph(x_constraints=[410, 1], edge_index=[2, 31510], edge_attr=[31510, 1], x_variables=[510, 9], num_nodes=920)
```
The `Batch` of the above instance looks like,
```
BipartiteGraphBatch(x_constraints=[590, 1], x_constraints_batch=[590], x_constraints_ptr=[3], edge_index=[2, 39640], edge_attr=[39640, 1], x_variables=[765, 9], x_variables_batch=[765], x_variables_ptr=[3], num_nodes=1355, batch=[1355], ptr=[3])

```

**I am trying to convert the sparse adjacency matrix to a dense adjacency matrix and finally reconstruct the sparse adjacency matrix using the recently converted dense adjacency matrix.** 

I have used the methods `to_dense_adj` and `dense_to_sparse` from `torch_geometric.utils` package. The conversion output seems to be a bit sketchy,

```
Sparse Shape: [2, 39640]
Dense Shape: [2, 920, 920]
Reconstructing Sparse with Above Dense Matrix: [2, 39566]
```

We can see **some edges are being missed in the newly constructed sparse adjacency matrix**. I didn't find the reason. Can anyone help me to fix the issue or give me the reason for the above issue? Thank you for reading and the help from the community will be much appreciated.

Custom `Data` function code is attached below,
```
class BipartiteGraph(Data):
    masked_cons_idx: int
    bias_label: FloatTensor
    degree_label: LongTensor
    logits_label: FloatTensor
    connected_vars_idx: LongTensor
    weights_label: FloatTensor

    def __init__(
        self,
        x_constraints: FloatTensor = None,
        edge_index: LongTensor = None,
        edge_attr: FloatTensor = None,
        x_variables: FloatTensor = None,
        **kwargs
    ):
        """
        Bipartite graph data object, each representing a MILP instance.
        """
        super().__init__(**kwargs)
        self.x_constraints = x_constraints
        self.edge_index = edge_index
        self.edge_attr = edge_attr
        self.x_variables = x_variables
        if x_constraints is not None:
            self.num_nodes = len(x_constraints) + len(x_variables)

    def __inc__(self, key, value, store, *args, **kwargs):
        if key == "edge_index":
            return torch.tensor(
                [[self.x_constraints.size(0)], [self.x_variables.size(0)]]
            )
        if key == "masked_cons_idx":
            return self.x_constraints.size(0)
        elif key == "connected_vars_idx":
            return self.x_variables.size(0)
        else:
            return super().__inc__(key, value, *args, **kwargs)

    @property
    def num_variables(self):
        return len(self.x_variables)

    @property
    def num_constraints(self):
        return len(self.x_constraints)
```

Again Thank You !!

**Comment by rusty1s** [comment]
I assume you hold some duplicated edges in `edge_index`, which should explain this. Can you confirm?

**Reply by srikesh-07** [reply]
Thank you for your suggestion @rusty1s !!!. 

I already tried to verify the edge indices of each instance in `Dataset` using the `coalesce` method. There are no duplicated edges,

Also, I am facing this reconstruction issue while using the `Batch` instance and not the single `Data` instance.

PS: Also tried with the `RemoveDuplicateEdges`, but no changes !!

**Reply by rusty1s** [reply]
Thanks for checking. Does `edge_index` hold the constraint of
```
assert edge_index[0].max() < x_constraints.size(0)
assert edge_index[1].max() < x_variables.size(0)
```
